src/order/filter/bases.lean

```diff
@@ -40,57 +63,204 @@ machinery, e.g., `simp only [exists_prop, true_and]` or `simp only [forall_const
+iff.rfl
+
+lemma mem_filter_of_mem (B : filter_basis α) {U : set α} : U ∈ B → U ∈ B.filter:=
+λ U_in, ⟨U, U_in, subset.refl _⟩
+
+lemma eq_infi_principal (B : filter_basis α) : B.filter = ⨅ s : B.sets, principal s :=
+begin
+  ext U,
+  rw [mem_filter_iff, mem_infi],
+  { simp },
+  { rintros ⟨U, U_in⟩ ⟨V, V_in⟩,
+    rcases B.inter_sets U_in V_in with ⟨W, W_in, W_sub⟩,
+    use [W, W_in],
+    finish },
+  cases B.nonempty with U U_in,
+  exact ⟨⟨U, U_in⟩⟩,
+end
+
+protected lemma generate (B : filter_basis α) : generate B.sets = B.filter :=
+begin
+  apply le_antisymm,
+  { intros U U_in,
+    rcases B.mem_filter_iff.mp U_in with ⟨V, V_in, h⟩,
+    exact generate_sets.superset (generate_sets.basic V_in) h },
+  { rw sets_iff_generate,
+    apply mem_filter_of_mem }
+end
+end filter_basis
+
+namespace filter
+namespace is_basis
+variables {p : ι → Prop} {s : ι → set α}
+
+/-- Constructs a filter from an indexed family of sets satisfying `is_basis`. -/
+protected def filter (h : is_basis p s) : filter α := h.filter_basis.filter
+
+protected lemma mem_filter_iff (h : is_basis p s) {U : set α} :
+  U ∈ h.filter ↔ ∃ i, p i ∧ s i ⊆ U :=
+begin
+  erw [h.filter_basis.mem_filter_iff],
+  simp only [mem_filter_basis_iff h, exists_prop],
+  split,
+  { rintros ⟨_, ⟨i, pi, rfl⟩, h⟩,
+    tauto },
+  { tauto }
+end
+
+lemma filter_eq_generate (h : is_basis p s) : h.filter = generate {U | ∃ i, p i ∧ s i = U} :=
+by erw h.filter_basis.generate ; refl
+end is_basis
 
 /-- We say that a filter `l` has a basis `s : ι → set α` bounded by `p : ι → Prop`,
 if `t ∈ l` if and only if `t` includes `s i` for some `i` such that `p i`. -/
-protected def has_basis (l : filter α) (p : ι → Prop) (s : ι → set α) : Prop :=
-∀ t : set α, t ∈ l ↔ ∃ i (hi : p i), s i ⊆ t
+protected structure has_basis (l : filter α) (p : ι → Prop) (s : ι → set α) : Prop :=
+(mem_iff' : ∀ (t : set α), t ∈ l ↔ ∃ i (hi : p i), s i ⊆ t)
 
 section same_type
 
 variables {l l' : filter α} {p : ι → Prop} {s : ι → set α} {t : set α} {i : ι}
   {p' : ι' → Prop} {s' : ι' → set α} {i' : ι'}
 
-/-- Definition of `has_basis` unfolded to make it useful for `rw` and `simp`. -/
+lemma has_basis_generate (s : set (set α)) : (generate s).has_basis (λ t, finite t ∧ t ⊆ s) (λ t, ⋂₀ t) :=
+⟨begin
+  intro U,
+  rw mem_generate_iff,
+  apply exists_congr,
+  tauto
+end⟩
+
+/-- The smallest filter basis containing a given collection of sets. -/
+def filter_basis.of_sets (s : set (set α)) : filter_basis α :=
+{ sets := sInter '' { t | finite t ∧ t ⊆ s},
+  nonempty := ⟨univ, ∅, ⟨⟨finite_empty, empty_subset s⟩, sInter_empty⟩⟩,
+  inter_sets := begin
+    rintros _ _ ⟨a, ⟨fina, suba⟩, rfl⟩ ⟨b, ⟨finb, subb⟩, rfl⟩,
+    exact ⟨⋂₀ (a ∪ b), mem_image_of_mem _ ⟨finite_union fina finb, union_subset suba subb⟩,
+           by rw sInter_union⟩,
+  end }
+
+/-- Definition of `has_basis` unfolded with implicit set argument. -/
 lemma has_basis.mem_iff (hl : l.has_basis p s) : t ∈ l ↔ ∃ i (hi : p i), s i ⊆ t :=
```

Comment:
Why do you have `mem_iff'` in the structure and `mem_iff` here? Even if the binders are different, I think I would directly put `mem_iff` in the structure (adjusting the binders), unless you have a good reason not to (and then some comment in the definition would be helpful).

Reply:
The issue is we need both versions.

Reply:
The only difference is the binders, right? I don't see where you use the prime version in a way that the non-primed version would not solve also.

Reply:
Sorry, I still don't see how to do that. In the structure I need the filter `l` to be explicit. When using the condition I want the `has_basis l p s` to be the explicit argument.